### DISCUSSION TITLE
Using static styles in templated control works incorrectly

### DISCUSSION THREAD

**Discussion by kucint** [unoplatform/uno.extensions#2349] [Q&A] [answered]
This issue is a continuation of #2316.

**Repro steps:**
Create a simple control with two predefined **static** styles: `MyStyleA` and `MyStyleB`
Assign `MyStyleA` to your control's style.

--------------------------
**Issue 1:**

this issue can be observed on windows:
TargetFramework: `net8.0-desktop;`

**Expected behavior**: `MyStyleA` is used, 
**Actual behavior**: `MyStyleB` is used, 

```cs
internal sealed class MyControl : Control
{
    public static readonly Style<MyControl> _styleA = new MyStyleA();   // red
    public static readonly Style<MyControl> _styleB = new MyStyleB();   // blue

    public MyControl()
    {
        DefaultStyleKey = typeof(MyControl);

        Style = _styleA;              // uses style B: error, expected style A
        //Style = _styleB;            // uses style B
        //Style = new MyStyleA();     // uses style A
        //Style = new MyStyleB();     // uses style B
    }
}

internal class MyStyleA : Style<MyControl>
{
    public MyStyleA() => Setters(s => s
        .Template(ctrl =>
            new Frame()
            .Background(Colors.LightCoral)
        ));
}

internal class MyStyleB : Style<MyControl>
{
    public MyStyleB() => Setters(s => s
        .Template(ctrl =>
            new Frame()
            .Background(Colors.LightSkyBlue)
        ));
}

public sealed partial class MainPage : Page
{
    public MainPage()
    {
        this
            .Background(ThemeResource.Get<Brush>("ApplicationPageBackgroundThemeBrush"))
            .Content(new StackPanel()
            .VerticalAlignment(VerticalAlignment.Center)
            .HorizontalAlignment(HorizontalAlignment.Center)
            .Children(
                new TextBlock()
                .Text("Hello Uno Platform!"),
                new MyControl()
                .Width(200)
                .Height(200)
            ));
    }
}
```
--------------------------
**Issue 2:**

this issue can be observed on windows:
TargetFramework: `net8.0-windows10.0.19041`

Application is unstable:
One of following may happen:
- app starts without errors but `MyControl` is not presented, terminates with error code = 0
- app starts without errors but `MyControl` is not presented, terminates with error code = 0xc0000374
- app fail to start throwing Exception = _{"No installed components were detected.\r\n\r\nCannot apply a Style with TargetType 'Microsoft.UI.Xaml.Controls.ControlTemplate' to an object of type 'UnoChangingControlStyleApp.Controls.MyControl'."}_

See #2311 for similar behavior

--------------------------
`"Uno.Sdk": "5.2.161"`
repro: [UnoChangingControlStyleApp.zip](https://github.com/user-attachments/files/15823207/UnoChangingControlStyleApp.zip)

**Comment by dr1rrb** [comment]
For the issue 1, I'm able to repro. I would suspect that we are registering both `Style` as implicit so the last one wins.

For issue 2, what I'm able to repro:
* App start without any exception, control not visible -- and `OnApplyTemplate` is not being invoked --
* *If I add a breakpoint in the ctor of the control* get an exception and app terminates:
```
Aucun composant installé n’a été détecté.

Cannot apply a Style with TargetType 'Microsoft.UI.Xaml.Controls.ControlTemplate' to an object of type 'UnoChangingControlStyleApp.Controls.MyControl'.
```

Definitely looks like a concurrency issue ...

Transferring to @dansiegel as master of C# markup

**Reply by MartinZikmund** [reply]
@kucint the last issue in step 2 seems the same as we discussed in https://github.com/unoplatform/uno.extensions/discussions/2316 . Have you tried adding the default .xaml style for the control? Maybe that will fix all of behaviors you observed in issue 2

**Reply by MartinZikmund** [reply]
For issue 1 we can investigate what the behavior should be, but most likely styles should be instead defined elsewhere (and be named) to make it more reliable, especially as static field initialization may not be relied upon. @dansiegel can you please link a sample of the recommended way to use control styles in C# markup-based apps?

**Reply by kucint** [reply]
> @kucint the last issue in step 2 seems the same as we discussed in #2316 . Have you tried adding the default .xaml style for the control? Maybe that will fix all of behaviors you observed in issue 2

sure I did :)

**Reply by dansiegel** [reply]
@kucint sorry for the late reply on this. I thought this was converted over to an issue already. I think this will be taken care of as part of the upcoming fix we have for #2405

**Comment by kucint** [comment]
Hi @dansiegel, the #2405 is already merged so I downloaded the newest uno sdk and tested it again.
Unfortunately, the issue persists on `"Uno.Sdk": "5.4.0-dev.145"`

**Reply by MartinZikmund** [reply]
Reopening for further investigation

**Reply by kucint** [reply, marked as answer]
this has been clarified in [#2399](https://github.com/unoplatform/uno.extensions/issues/2399), all is OK.
